[Clang][SME2] Add outer product and accumulate/subtract builtins

kmclaughlin-arm opened the pull request:
Adds the following SME2 builtins:
 - svmop(a|s)_za32,
 - svbmop(a|s)_za32

See https://github.com/ARM-software/acle/pull/217

kmclaughlin-arm committed:
[Clang][SME2] Add outer product and accumulate/subtract builtins

Adds the following SME2 builtins:
 - svmop(a|s)_za32,
 - svbmop(a|s)_za32

See https://github.com/ARM-software/acle/pull/217

kmclaughlin-arm committed:
- Add arm_streaming & arm_shared_za attributes to the tests

MDevereau commented on clang/test/CodeGen/aarch64-sme2-intrinsics/acle_sme2_mop.c:
```diff
@@ -0,0 +1,170 @@
+// CPP-CHECK-NEXT:    tail call void @llvm.aarch64.sme.bmopa.za32.nxv4i32(i32 3, <vscale x 4 x i1> [[TMP0]], <vscale x 4 x i1> [[TMP1]], <vscale x 4 x i32> [[ZN:%.*]], <vscale x 4 x i32> [[ZM:%.*]])
+// CPP-CHECK-NEXT:    ret void
+//
+void test_svbmopa_u16(svbool_t pn, svbool_t pm, svuint32_t zn, svuint32_t zm) __arm_streaming __arm_shared_za {
```
All of the svbmopa/svbmops test names have the wrong suffix.

Given BMOPA/BMOPS is a different operation from MOPA/MOPS, would these be better separated into `acle_sme2_bmop.c`?

kmclaughlin-arm replied:
Thanks, I've fixed the test names and split the bmopa/s tests into their own file.

MDevereau commented on clang/test/Sema/aarch64-sme2-intrinsics/acle_sme2_imm.cpp:
```diff
@@ -0,0 +1,18 @@
+  svmops_za32_s16_m(4, pred, pred, s16, s16); // expected-error {{argument value 4 is outside the valid range [0, 3]}}
+
+  svbmopa_za32_u32_m(4, pred, pred, u32, u32); // expected-error {{argument value 4 is outside the valid range [0, 3]}}
+  svbmops_za32_s32_m(4, pred, pred, s32, s32); // expected-error {{argument value 4 is outside the valid range [0, 3]}}
```
It looks like tests `svbmopa_za32_s32_m` and `svbmops_za32_u32_m` are missing.

MDevereau commented on clang/include/clang/Basic/arm_sme.td:
```diff
@@ -298,3 +298,19 @@ multiclass ZAAddSub<string n_suffix> {
+
+  def SVSMOPS  : Inst<"svmops_za32[_{d}]_m", "viPPdd", "s", MergeNone, "aarch64_sme_smops_za32", [IsSharedZA, IsStreaming], [ImmCheck<0, ImmCheck0_3>]>;
+  def SVUSMOPS : Inst<"svmops_za32[_{d}]_m", "viPPdd", "Us", MergeNone, "aarch64_sme_umops_za32", [IsSharedZA, IsStreaming], [ImmCheck<0, ImmCheck0_3>]>;
+
```
```suggestion
//
//  Bitwise exclusive NOR population count outer product and accumulate/subtract
//
```

kmclaughlin-arm replied:
I think the "Outer product and accumulate/subtract" header applies to all of these, including the bmopa/s builtins? I've fixed the typo though :)

kmclaughlin-arm committed:
- Fixed typo in arm_sme.td comment
- Fixed bmopa/s test names and split into their own file
- Added missing tests to acle_sme2_imm.cpp
